Add Kandinsky 2.1

yiyixuxu opened the pull request:
this PR add Kandinsky2.1 to diffusers

https://github.com/huggingface/diffusers/issues/2985

original codebase: https://github.com/ai-forever/Kandinsky-2

to-do:
- [x] add image prior diffusion (with `prior_tokenizer`, `prior_text_encoder`, `prior_scheduler`)
- [x] text_proj
- [x] add unet
- [x] add text-to-image diffusion pipeline
- [x] add `image_encoder`, `text_encoder`, `tokenizer` 
- [x] add inpainting pipeline
- [x] MoVQ decoder 
- [x] MoVQ encoder 
- [x] tests 
- [x] docs
- [x] add img2img pipeline https://github.com/huggingface/diffusers/pull/3426

```python
from diffusers import KandinskyPipeline, KandinskyPriorPipeline

import torch
import numpy as np
device = "cuda"

# # inputs
prompt= "red cat, 4k photo"
batch_size=1 


# # create prior 
pipe_prior = KandinskyPriorPipeline.from_pretrained("YiYiXu/Kandinsky-prior")
pipe_prior.to("cuda")

# use prior to generate image_emb based on our prompt
generator = torch.Generator(device=device).manual_seed(0)
out = pipe_prior(prompt, generator=generator,)
image_emb = out.images
zero_image_emb = out.zero_embeds

# create diffuser pipeline
pipe = KandinskyPipeline.from_pretrained("YiYiXu/Kandinsky")
pipe.to(device)


generator = torch.Generator(device="cuda").manual_seed(0)
out = pipe(
    prompt,
    image_embeds=image_emb,
    negative_image_embeds =zero_image_emb,
    height=768,
    width=768,
    num_inference_steps=100,
    generator=generator )

image = out.images[0]
image.save("cat.png")
```

![yiyi_test_pipe_kandinsky5_out_new](https://github.com/huggingface/diffusers/assets/12631849/32f4e2cf-79b0-492c-a099-0d1e185df33a)


use inpainting pipeline to add a hat 

```python
from diffusers import KandinskyInpaintPipeline, KandinskyPriorPipeline
from diffusers.utils import load_image

import torch
import numpy as np

device = "cuda"

REPO_PRIOR = "YiYiXu/Kandinsky-prior"
REPO_INPAINT = "YiYiXu/Kandinsky-inpaint"

# # inputs
prompt= "a hat"
batch_size=1 

# # create prior 
pipe_prior = KandinskyPriorPipeline.from_pretrained(REPO_PRIOR)
pipe_prior.to("cuda")

# use prior to generate image_emb based on our prompt
generator = torch.Generator(device=device).manual_seed(0)
out = pipe_prior(prompt, generator=generator,)
image_emb = out.images
zero_image_emb = out.zero_embeds

# create diffuser pipeline
pipe = KandinskyInpaintPipeline.from_pretrained(REPO_INPAINT )
pipe.to(device)

init_image = load_image(
    "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main" 
    "/kandinsky/cat.png")

mask = np.ones((768, 768), dtype=np.float32)
mask[:250,250:-250] =  0

generator = torch.Generator(device="cuda").manual_seed(0)
out = pipe(
    prompt,
    image=init_image,
    mask_image=mask,
    image_embeds=image_emb,
    negative_image_embeds =zero_image_emb,
    height=768,
    width=768,
    num_inference_steps=150,
    generator=generator )

image = out.images[0]
image.save("cat_with_hat.png")

```


![cat_with_hat](https://github.com/huggingface/diffusers/assets/12631849/587f3ea2-8922-4bf2-919b-fddbf873e0f9)

image-to-image generation 
```python
from diffusers.utils import load_image
from diffusers import KandinskyImg2ImgPipeline, KandinskyPriorPipeline, DDIMScheduler
import torch

REPO_PRIOR = "YiYiXu/Kandinsky-prior"
REPO = "YiYiXu/Kandinsky"

model_dtype = torch.float16

prompt = "A red cartoon frog, 4k"

init_image = load_image(
    "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main" 
    "/kandinsky/frog.png")

# create prior 
pipe_prior = KandinskyPriorPipeline.from_pretrained(REPO_PRIOR, torch_dtype=model_dtype)
pipe_prior.to("cuda")

# use prior to generate image_emb based on our prompt
generator = torch.Generator(device='cuda').manual_seed(0)
out = pipe_prior(prompt, num_inference_steps=25, generator=generator,).
image_emb = out.images
zero_image_emb = out.zero_embeds

# create img2img pipeline
pipe = KandinskyImg2ImgPipeline.from_pretrained(REPO, torch_dtype=model_dtype)

# create ddim scheduler 
ddim_config = {
    "num_train_timesteps": 1000,
    "beta_schedule":  "linear",
    "beta_start": 0.00085,
    "beta_end":0.012,
    "clip_sample" : False,
    "set_alpha_to_one" : False, 
    "steps_offset" : 0,
    "prediction_type" : "epsilon",
    "thresholding" : False,
}

ddim_scheduler = DDIMScheduler(**ddim_config)
pipe.scheduler = ddim_scheduler

pipe.to("cuda")


generator = torch.Generator(device='cuda').manual_seed(0)
out = pipe(
    prompt=prompt, 
    image=init_image, 
    height=768, 
    width=768, 
    num_inference_steps=100, 
    generator=generator, 
    image_embeds=image_emb, 
    negative_image_embeds=zero_image_emb, 
    strength=0.2,) 

out.images[0].save("img2img_frog_out.png")

```

![frog](https://github.com/huggingface/diffusers/assets/12631849/1592aef8-03f3-47cb-adce-1fbf4bc75477)

![yiyi_test_img2img1_out](https://github.com/huggingface/diffusers/assets/12631849/5f495db6-a85e-4a70-b3b5-d8974f512454)

image mixing 

```python
from diffusers import KandinskyPriorPipeline, KandinskyPipeline
from diffusers.utils import load_image
import PIL

import torch

from torchvision import transforms

REPO_PRIOR = "YiYiXu/Kandinsky-prior"
REPO = "YiYiXu/Kandinsky"

model_dtype = torch.float16


# create prior 
pipe_prior = KandinskyPriorPipeline.from_pretrained(REPO_PRIOR, torch_dtype=model_dtype)
pipe_prior.to("cuda")

# we will use prior to create image_emb and zero_image_emb

img1 = load_image("https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main" 
    "/kandinsky/cat.png")

img2 = load_image("https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main" 
    "/kandinsky/starry_night.jpeg")

images_texts = ["a cat", img1, img2 ]
weights = [0.3,0.3,0.4]

image_emb, zero_image_emb = pipe_prior.interpolate(images_texts, weights,)

pipe = KandinskyPipeline.from_pretrained(REPO, torch_dtype =model_dtype )
pipe.to("cuda")

generator = torch.Generator(device="cuda").manual_seed(0)
out = pipe(
    "",
    image_embeds=image_emb,
    negative_image_embeds =zero_image_emb,
    height=768,
    width=768,
    num_inference_steps=150,
    generator=generator )

image = out.images[0]
image.save("starry_cat.png")
```

![starry_cat](https://github.com/huggingface/diffusers/assets/12631849/0245d3bd-5ba6-4163-8bbf-2f85755dd9ce)

isamu-isozaki commented:
@yiyixuxu Very cool! Just thought to mention but we did port MoVQ in huggingface/open-muse [here](https://github.com/huggingface/open-muse) which might help

yiyixuxu commented:
@isamu-isozaki thanks!

ayushtues commented:
Very cool @yiyixuxu, can you tell me how you tested if the prior model is working and if the weights were loading?  Would it be cool to have a temporary jupyter-notebook handy for testing the pipeline, and its individual components, if the weights are loading etc

ayushtues commented:
For the decoder, it is practically the same as the VQVAE model already in diffusers https://github.com/huggingface/diffusers/blob/kandinsky/src/diffusers/models/vq_model.py, but it just uses a different custom normalization layer in the decoder (SpatialNorm) which takes an extra embedding as input, than GroupNorm,  and thus needs changes to the attention/resnet building blocks, which are also the same as the ones present in diffusers, except the normalization layer (they use Groupnorm, and need to use SpatialNorm now).

We can either parametrize the attention/resnet building blocks and VQVAE in diffusers to support using a different normalization layer, or copy them with the minimal changes in the Kandinskt pipeline if we feel the normalization layer is not general enough to change the existing implementations

yiyixuxu commented:
@ayushtues let's add SpatialNorm to the blocks

yiyixuxu commented:
@ayushtues this is the script I use to do a quick compare along the process

note that this might not work for you because I had to go into the original repo to hardcode a few things to make sure we can reproduce ( the noise construction is a little bit different in diffusers, I don't think you will need to do this for decoder) - but you can use a similar process 

```python
import numpy as np
from kandinsky2 import get_kandinsky2
import torch
model = get_kandinsky2('cuda', task_type='text2img', model_version='2.1', use_flash_attention=False)

prompt= "red cat, 4k photo"
batch_size=1 
guidance_scale=4
prior_cf_scale=4,
prior_steps="5"
negative_prior_prompt=""

# generate clip embeddings
image_emb = model.generate_clip_emb(
    prompt,
    batch_size=batch_size,
    prior_cf_scale=prior_cf_scale,
    prior_steps=prior_steps,
    negative_prior_prompt=negative_prior_prompt,
        )
print(f"image_emb:{image_emb.shape},{image_emb.sum()}")




# diffusers
from diffusers import KandinskyPipeline, PriorTransformer
import diffusers

from diffusers.training_utils import set_seed

prompt= "red cat, 4k photo"
device = "cuda"

pipe_prior = KandinskyPipeline.from_pretrained("/home/yiyi_huggingface_co/test-kandinsky")
pipe_prior.to(device)

generator = torch.Generator(device="cuda").manual_seed(0)
image_emb_d = pipe_prior(
    prompt,
    generator=generator,
)

print(f"image_embeddings:{image_emb_d.shape},{image_emb_d.sum()}")
print("compare results:")
print(np.max(np.abs(image_emb_d.detach().cpu().numpy() - image_emb.detach().cpu().numpy())))
```

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_3308). All of your documentation changes will be reflected on that endpoint.

ayushtues commented:
Okay the outputs are in 1e-4 of each other for the decoder and 1e-5 for the encoder and seem similar, so should be okay

yiyixuxu commented:
@ayushtues 

Thanks for adding the decoder so fast! super awesome job! 😇🤗👏👍 

I think we can wrap up Kandinsky soon! A few tasks left (I ranked them from easy to difficult based on my subjective judgment 😂)  - let me know if you are interested in taking any of these.  I will help you as much as you need of course :) 

1. image encoder: I think it's only used to generate zero image embedding for unconditional tokens here https://github.com/ai-forever/Kandinsky-2/blob/main/kandinsky2/kandinsky2_1_model.py#L323, it's just the clip image encoder
```python
from transformers import CLIPVisionModelWithProjection
clip_image_encoder = CLIPVisionModelWithProjection.from_pretrained("openai/clip-vit-large-patch14", torch_dtype=torch.float16).to("cuda")
```
2. second `text_encoder` and `tokenizer` for the text-to-image diffusion process with Unet 
    -  note that it's different from the `prior_text_encoder` and `prior_tokenizer` (CLIP text_encoder and tokenizer) - they use `MultilingualCLIP` for text_to_image
    -  `MultilingualCLIP` is not in `transformers` library yet but it's just a few lines of code, we can just take this https://github.com/ai-forever/Kandinsky-2/blob/main/kandinsky2/model/text_encoders.py#L108
4. second scheduler for text-to-image diffusion process: I think we can just use `UnCLIPScheduler` https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_unclip.py out-of-box with:
    - `variance_type : "learned_range"`
    - `clip_sample_range:2`
    -  we may need to add this bit to `UnCLIPScheduler` https://github.com/ai-forever/Kandinsky-2/blob/main/kandinsky2/model/gaussian_diffusion.py#L287
5. put everything together into the pipeline!

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/text_proj.py:
```diff
@@ -0,0 +1,77 @@
+# Copyright 2023 The HuggingFace Team. All rights reserved.
```
Cool!

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py:
```diff
@@ -0,0 +1,269 @@
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+class KandinskyPipeline(DiffusionPipeline):
```
Looks very nice so far!

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py:
```diff
@@ -0,0 +1,269 @@
+                prev_timestep=prev_timestep,
+            ).prev_sample
+
+        prior_latents = self.prior.post_process_latents(prior_latents)
```
Great to directly us the new processor here!

ayushtues commented:
@yiyixuxu I can take up 1, 2, and then later help with 4 when the parts are ready to combine in the pipeline; not so familiar with how schedulers integrate into diffusers, so will leave 3 to you, but will definitely want to review it and learn how they integrate into the pipeline.

yiyixuxu commented:
@ayushtues great!

ayushtues commented:
@yiyixuxu where do you think we should put the multilingualCLIP model, since it's not directly available in HF, should we add it in a separate file in `pipelines/Kandinsky`? 

Also it does not have a `save_pretrained` function, so it cannot be directly loaded using `from_pretrained`, but instead needs to be passed a path to a config file given in the Kandinsky repo containing the weights and config, so we will need to copy those weights over in our repo as well, does that sound okay?

ayushtues commented:
Meanwhile started another PR for task 1, 2 - #3373

patrickvonplaten commented on src/diffusers/schedulers/scheduling_unclip.py:
```diff
@@ -97,6 +97,15 @@ class UnCLIPScheduler(SchedulerMixin, ConfigMixin):
         prediction_type (`str`, default `epsilon`, optional):
```
Ok for me!

yiyixuxu commented:
```python
from kandinsky2.utils import process_images
from kandinsky2.model.text_encoders import TextEncoder
from kandinsky2.vqgan.autoencoder import MOVQ

from diffusers import KandinskyPipeline, KandinskyPriorPipeline

import clip
from transformers import AutoTokenizer

import torch
import numpy as np

device = "cuda"

# # inputs
prompt= "red cat, 4k photo"
batch_size=1 
guidance_scale=4
prior_cf_scale=4,
prior_steps="5"
negative_prior_prompt = ""

# # create prior using KandinskyPriorPipeline
pipe_prior = KandinskyPriorPipeline.from_pretrained("/home/yiyi_huggingface_co/model_repo/Kandinsky-prior")
pipe_prior.to("cuda")

# use prior to generate image_emb based on our prompt
generator = torch.Generator(device=device).manual_seed(0)
image_emb = pipe_prior(
    prompt,
    generator=generator,
)

# use clip image_encoder to create zero_image_emb (TO-DO: need to use transformers) 
clip_model, _ = clip.load("ViT-L/14", device=device, jit=False)
zero_image_emb = torch.zeros(1, 3, 224, 224, dtype= torch.float16).to(device)
zero_image_emb = clip_model.encode_image(zero_image_emb).to(dtype=torch.float16)
zero_image_emb = zero_image_emb.repeat(batch_size,1)

# use M-CLIP text_encoder and tokenizer to process prompt (TO-DO: need to port to diffusers)
# create text_encoder 
text_encoder_config = {'model_path': '/tmp/kandinsky2/2_1/text_encoder', 'model_name': 'multiclip', 'in_features': 1024, 'out_features': 768}
text_encoder = TextEncoder(**text_encoder_config)
text_encoder = text_encoder.half().to(device).eval()

# create tokenizer
tokenizer1 = AutoTokenizer.from_pretrained('/tmp/kandinsky2/2_1/text_encoder')

# create prompt_embeds and text_encoder_hidden_states
text_encoding = tokenizer1(
            [prompt] * batch_size + [""] * batch_size,
            max_length=77,
            padding="max_length",
            truncation=True,
            return_attention_mask=True,
            add_special_tokens=True,
            return_tensors="pt",
        )

tokens = text_encoding["input_ids"].to(device)
mask = text_encoding["attention_mask"].to(device)

full_emb, pooled_emb = text_encoder(tokens=tokens, mask=mask)

# create diffuser pipeline
local_dir = "/home/yiyi_huggingface_co/model_repo/Kandinsky"
model_dtype = torch.float32
pipe = KandinskyPipeline.from_pretrained(local_dir, torch_dtype =model_dtype )
pipe.to(device)

# [cond, uncond] -> [uncond, cond]
pooled_emb = torch.flip(pooled_emb, dims=(0,)).to(model_dtype)
full_emb = torch.flip(full_emb, dims=(0,)).to(model_dtype)
image_emb = image_emb.to(model_dtype)
zero_image_emb = zero_image_emb.to(model_dtype)

generator = torch.Generator(device="cuda").manual_seed(0)
samples = pipe(
    prompt_embeds=pooled_emb,
    text_encoder_hidden_states=full_emb,
    image_embeds=image_emb,
    negative_image_embeds =zero_image_emb,
    height=768,
    width=768,
    num_inference_steps=100,
    generator=generator )

# use MOVQ decoder to decode generated image (TO-DO: need to add this to diffusers)
# create decoder 
movq_config = {
    'embed_dim': 4, 
    'n_embed': 16384, 
    'ddconfig': {
        'double_z': False, 
        'z_channels': 4, 
        'resolution': 256, 
        'in_channels': 3, 
        'out_ch': 3, 
        'ch': 128, 
        'ch_mult': [1, 2, 2, 4], 
        'num_res_blocks': 2, 
        'attn_resolutions': [32], 
        'dropout': 0.0}
        }

movq_ckpt_path = "/tmp/kandinsky2/2_1/movq_final.ckpt"

image_encoder = MOVQ(**movq_config)
image_encoder.load_state_dict(
                    torch.load(movq_ckpt_path)
                )
image_encoder.half().to("cuda").eval()

samples = samples.to(dtype=torch.float16)
samples = image_encoder.decode(samples)          
samples = samples[:, :, :768, :768]
images_d = process_images(samples)
images_d[0].save("k_image_d_test.png")
```


![k_image_d_test](https://github.com/huggingface/diffusers/assets/12631849/ca74c7c4-8987-4cd5-bce7-55458f4d63cd)

patrickvonplaten commented:
@yiyixuxu let me know if you need a final review :-)

yiyixuxu commented:
ready now @patrickvonplaten

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/__init__.py:
```diff
@@ -0,0 +1,18 @@
+
+
+try:
+    if not (is_transformers_available() and is_torch_available() and is_transformers_version(">=", "4.25.0")):
```
```suggestion
    if not (is_transformers_available() and is_torch_available()):
```
>= 4.25.1 is a necessary dependency

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,347 @@
+    Args:
+        prior ([`PriorTransformer`]):
+            The canonincal unCLIP prior to approximate the image embedding from the text embedding.
+        prior_image_encoder ([`CLIPVisionModelWithProjection`]):
```
```suggestion
        image_encoder ([`CLIPVisionModelWithProjection`]):
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,347 @@
+            The canonincal unCLIP prior to approximate the image embedding from the text embedding.
+        prior_image_encoder ([`CLIPVisionModelWithProjection`]):
+            Frozen image-encoder.
+        prior_text_encoder ([`CLIPTextModelWithProjection`]):
```
```suggestion
        text_encoder ([`CLIPTextModelWithProjection`]):
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,347 @@
+            Frozen image-encoder.
+        prior_text_encoder ([`CLIPTextModelWithProjection`]):
+            Frozen text-encoder.
+        prior_tokenizer (`CLIPTokenizer`):
```
```suggestion
        tokenizer (`CLIPTokenizer`):
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,347 @@
+        prior_tokenizer (`CLIPTokenizer`):
+            Tokenizer of class
+            [CLIPTokenizer](https://huggingface.co/docs/transformers/v4.21.0/en/model_doc/clip#transformers.CLIPTokenizer).
+        prior_scheduler ([`UnCLIPScheduler`]):
```
```suggestion
        scheduler ([`UnCLIPScheduler`]):
```

patrickvonplaten replied:
Now that the pipelines are split, let's make the classes name easy

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,347 @@
+    def __init__(
+        self,
+        prior: PriorTransformer,
+        prior_image_encoder: CLIPVisionModelWithProjection,
+        prior_text_encoder: CLIPTextModelWithProjection,
+        prior_tokenizer: CLIPTokenizer,
+        prior_scheduler: UnCLIPScheduler,
```
```suggestion
        image_encoder: CLIPVisionModelWithProjection,
        text_encoder: CLIPTextModelWithProjection,
        tokenizer: CLIPTokenizer,
        scheduler: UnCLIPScheduler,
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,347 @@
+            prior_text_encoder=prior_text_encoder,
+            prior_tokenizer=prior_tokenizer,
+            prior_scheduler=prior_scheduler,
+            prior_image_encoder=prior_image_encoder,
```
```suggestion
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/text_proj.py:
```diff
@@ -0,0 +1,71 @@
+from ...models import ModelMixin
+
+
+class KandinskyTextProjModel(ModelMixin, ConfigMixin):
```
Actually I think we should move all of this directly into the UNet as this will make it much easier to fine-tune Kadinsky. @yiyixuxu just leave it for now, happy to help you do the change in the very end :-)

patrickvonplaten commented:
This PR is already in a great state! A couple thoughts here regarding the **design**:

# 1. Design change compared to [Karlo](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/unclip/pipeline_unclip.py) or [SD-UnCLIP](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_unclip.py)

As done in this PR we're creating two pipelines instead of one. One pipeline is the **pior** pipeline that takes care of diffusing the text embeddings to image embeddings and the other pipeline is the **decoder** pipeline that takes care of diffusing noise conditioned on image embeddings to images. This is in contrast to Karlo and UnCLIP above, but I think it's the right move because:
- This way every pipeline only has denoising process which is much easier to understand for the user
- We can have better naming and better split the pipelines into different repos. E.g. it might very well be that future work will only build on top of the pior or the decoder and in this case it's better to have them split
- It's easier to deal with memory management
- It's the right approach to make pipelines more atomic - we move from one pipeline rules it all to a "chaining" approach
- Better split of responsibility 

The disadvantage here is that a) the user **has to** run two pipelines to get the pipeline to work (but that's already the case with IF anyways) and b) that it stands in contrast to how Karlo and SD-UnCLIP are build. However those pipelines are really not used much, therefore I think it's fine to go with this approach. **Also** we could in the future refactor Karlo and SD-UnCLIP to follow this approach if necessary.

=> The design principle is quite easy to follow here:
- the **prior** takes care of everything -> image_embedding and the **decoder** takes care of (noise, image_embedding) -> image
- The prior pipeline always stays the same for all applications, only the decoder pipeline needs three variants: img2img, text2img, inpainting

@williamberman @sayakpaul @pcuenca ok for you?

# 2. We should make sure to support all use cases here:

1.) Text2Img
Already works! 
2.) Inpainting & Img2Img
Should be easy to add once text2img works
3.) Image mixing - that's quite a cool application and Kadinsky seems to be a good fit here. For this we could actually just use transformers' Vision CLIP to encode the image and then just as before we can use the `text2img` pipeline, no need for bigger changes here

@yiyixuxu could you maybe have a look into the other use cases besides 1.) as well and add them?

# 3. As discussed with @williamberman I think we should move away from the design pattern of creating stand-alone "text-encoder -> to -> unet" link/mapping models (e.g. [those](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/unclip/text_proj.py) ones) that take care of mapping text encoder and other other model's output (here it'd be also the image embedddings). 

We've used this design pattern for Karlo and SD-UnCLIP, but **not** for sdx4 or IF. For SDx4 and IF we've moved those layers directly into the UNet. While it's really a tricky decision because both approaches have pro's and con's I'm in favor of moving everything to the UNet mainly because I see them "weights that adapt the text encoder outputs for the unet" and thus to me they should be in the unet. This has a couple of advantages:
- a.) For fine-tuning one always only needs a text encoder and a unet and doesn't have to think about text encoder, mapping model and unet. While the text encoder might very well stay frozen for training, the mapping model and unet are **always** trained together, so let's just move the mapping model into the unet in which case the user **only** always needs to think about an unet.
- b.) The mapping models often project the text embeddings and other embedding to time step embeddings which are then added at every step. Since we always have the 1 -> time_proj -> time_emb weights in the unet, we should also have the <condition> -> condition_proj -> added_time_emb in the unet as it's cleaner this way. It cleaner IMO to have:
```py
def forward(a, b):
    a' = self.proj(a)
    b' = self.proj(b)
    out = a' + b'
```
then:
```py
def forward(a, b'):
     a' = self.proj(a)
     out = a' + b'
```
- c.) By moving everything into the UNet we give the user more freedom to create a large amount of different unets for pretraining. This however requires us to have clean docstrings an clean code in the UNet..

The disadvantage of this design choice is that:
d) The UNet2DConditionModel will become harder to read in the beginning. I think we can remedy this however by a nice refactor
e) It's a tiny bit inefficient to always run the same computation in every loop. E.g. if we down-project the text embeddings to a lower dimension we could have probably done this **before** the loop as the values don't change. However in practice this will never make more than 1% speed difference since the condition projects are tiny compared to the whole unet.

Very curios to hear your thoughts @sayakpaul @williamberman @yiyixuxu @pcuenca 

For this PR @yiyixuxu please leave everything as is for now. I'm more than happy to help you before merging the PR to refactor the text proj class

# 4. Let's think about some cool use cases for advertisement. To me it seems the model is very good at creating very high quality images so it might make sense to think a bit about some cool demos, whether this model could be nicely dreamboothed or not, etc...

patrickvonplaten commented:
@yiyixuxu sorry for the long message. In terms of next actionable steps, it'd be great to:
a) Move the MoVQ decoder into the kadinsky pipeline
b) Add img2img and inpainting
c) Upload two different pipeline repos (we'll have one prior and one kadinsky repo)
d) tests and documentation 

Let me know if something is not clear :-)

sayakpaul commented:
> 1. Design change compared to Karlo or SD-UnCLIP

Agreed. 

> 3. Move away from "mapping" model design pattern

Not in agreement as what you're proposing makes:

* Different things entangled for a single model
* The code bloated and likely very hard to read (it's already a bit hard to navigate through) 

During fine-tuning, if we force the users to be aware of the separate modules, IMO, that's better than having them forget about freezing a part of the network. 

>  It's a tiny bit inefficient to always run the same computation in every loop. E.g. if we down-project the text embeddings to a lower dimension we could have probably done this before the loop as the values don't change. However in practice this will never make more than 1% speed difference since the condition projects are tiny compared to the whole unet.

Which loop?

patrickvonplaten commented:
> Not in agreement as what you're proposing makes:
> 
> Different things entangled for a single model
> The code bloated and likely very hard to read (it's already a bit hard to navigate through)
> During fine-tuning, if we force the users to be aware of the separate modules, IMO, that's better than having them forget about freezing a part of the network.

Thanks for the answer, good to discuss this here! Please also share your thoughts here @pcuenca @williamberman @yiyixuxu @patil-suraj 

Some follow-up questions:
> Different things entangled for a single model
Why are they different things to you (the text emb, img emb -> time emb mapping and the unet?). To me that's all part of the unet?

> The code bloated and likely very hard to read (it's already a bit hard to navigate through)
Yes, I fully understand this argument! TBH, I don't think it would look much worse than it already is.

> During fine-tuning, if we force the users to be aware of the separate modules, IMO, that's better than having them forget about freezing a part of the network.
Not sure whether we're on the same page here. These weights should **not** be frozen. Whenever the model is fine-tuned all the mapping weights **have to be fine-tuned** as well. The idea here is exactly that the user don't have to think about anything and can just fine-tune the unet like all other unets. If we seperate mapping model and unet, the model almost becomes un-finetuneable because:
- All the checkpointing code needs to be adapted to in addition to the unet also save the mapping model
- We need to wrap accelerate around both modules
- People cannot share just a trained unet, they would in addition always have to share the mapping model

yiyixuxu commented:
>If we seperate mapping model and unet, the model almost becomes un-finetuneable because:
> 
> * All the checkpointing code needs to be adapted to in addition to the unet also save the mapping model
> * We need to wrap accelerate around both modules
> * People cannot share just a trained unet, they would in addition always have to share the mapping model


@patrickvonplaten 
would the process to fine-tune be similar to text-to-image while we train both text_encoder and unet?

sayakpaul commented:
> would the process to fine-tune be similar to text-to-image while we train both text_encoder and unet?

For text-to-image fine-tuning, it's very rare to fine-tune both, though. But I don't suspect any differences there. Did you have anything specific in mind? 

@patrickvonplaten I am convinced about the changes you proposed regarding the UNet. Your latest points drove me home. Thanks for elaborating!

yiyixuxu commented:
inpainting pipeline working 

MoVQ decoder is not in diffusers yet so we first create it from the original repo 
```python
from kandinsky2.vqgan.autoencoder import MOVQ
# create image_encoder 

movq_config = {
    'embed_dim': 4, 
    'n_embed': 16384, 
    'ddconfig': {
        'double_z': False, 
        'z_channels': 4, 
        'resolution': 256, 
        'in_channels': 3, 
        'out_ch': 3, 
        'ch': 128, 
        'ch_mult': [1, 2, 2, 4], 
        'num_res_blocks': 2, 
        'attn_resolutions': [32], 
        'dropout': 0.0}
        }

movq_ckpt_path = "/tmp/kandinsky2/2_1/movq_final.ckpt"

image_encoder = MOVQ(**movq_config)
image_encoder.load_state_dict(
                    torch.load(movq_ckpt_path)
                )
image_encoder.to("cuda").eval()
```

add a hat to the cat image we just generated with the Kandinsky text2img pipeline

```python
from diffusers import KandinskyInpaintPipeline, KandinskyPriorPipeline

import torch
import numpy as np

from PIL import Image
device = "cuda"

REPO_PRIOR = "YiYiXu/Kandinsky-prior"
REPO_INPAINT = "YiYiXu/Kandinsky-inpaint"

# # inputs
prompt= "a hat"
batch_size=1 

# # create prior 
pipe_prior = KandinskyPriorPipeline.from_pretrained(REPO_PRIOR)
pipe_prior.to("cuda")

# use prior to generate image_emb based on our prompt
generator = torch.Generator(device=device).manual_seed(0)
image_emb = pipe_prior(prompt, generator=generator,)
zero_image_emb = pipe_prior("")

# create inpaint pipeline
pipe = KandinskyInpaintPipeline.from_pretrained(REPO_INPAINT)
pipe.to(device)
# now we manually attached the decoder to the pipeline
pipe.image_encoder = image_encoder 

init_image = Image.open('cat.png')
mask = np.ones((768, 768), dtype=np.float32)
mask[:250,250:-250] =  0

generator = torch.Generator(device="cuda").manual_seed(0)
out = pipe(
    prompt,
    image=init_image,
    mask_image=mask,
    image_embeds=image_emb,
    negative_image_embeds =zero_image_emb,
    height=768,
    width=768,
    num_inference_steps=150,
    generator=generator )

image = out.images[0]
image.save("cat_with_hat.png")
```
![cat_with_hat](https://github.com/huggingface/diffusers/assets/12631849/27fd5564-63e3-46f5-a17c-403c74078dca)

ayushtues commented:
Awesome work @yiyixuxu! Do you want me to pick up adding the img2img pipeline?

williamberman commented:
I've previously discussed with patrick that my opinion is as much of the embedding (both time and text) preprocessing should be moved out of the unet body as possible. I think as is we're risking maintenance burden. I think it's reasonable to include the embedding pre-processing in the unet _only_ if that embedding pre-processing is sufficiently simple which would probably require separate unet definitions for sufficiently different pipelines (I think this is closer to transformers design policy).

As is, I think it's better to keep things closer to the current design and include embedding processing in the unet since I think we've already committed to it.

williamberman commented:
@yiyixuxu this is great! 

One note is that we shouldn't use the unCLIP scheduler. I used that while implementing the unCLIP pipeline because there was a bug in the the DDPM scheduler that we didn't realize was a bug, and we wanted to configure the pipeline with separate timesteps instead of just updating the DDPM scheduler to work with both ways of computing timesteps. 

We fixed the bug in the DDPM scheduler and tbh I should just update the unCLIP pipeline to use it at some point. I'll add a comment on the unCLIP scheduler that it shouldn't be used so it isn't confusing in the future :)

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/text_encoder.py:
```diff
@@ -0,0 +1,33 @@
+from transformers import AutoModel, PretrainedConfig, PreTrainedModel
+
+
+class MCLIPConfig(PretrainedConfig):
```
```suggestion
class MCLIPConfig(XLMRobertaConfig):
```

patrickvonplaten commented:
@yiyixuxu let me know if you need help or a final review :-)

yiyixuxu commented:
@patrickvonplaten 

going to take a deeper look into img2img https://github.com/huggingface/diffusers/pull/3426 tomorrow - the PR looks pretty good, but I think the image generated from the original repo still got a lot more details in it. will try to achieve same results. 

The rest of the PR here is ready for a review

One note: @williamberman mentioned that we should use the DDPM scheduler instead, but it does not work out of the box - should we do the refactor now? or should we keep unclip scheduler separate for now?

yiyixuxu commented:
@patrickvonplaten 

PR is ready for review now - 
I'm playing with image mixing right now - it's fun:) I think we (probably) don't need a pipeline for that - just use prior and text2image. so maybe I can just write an example (this model is so fun that you can feed it texts and images in so many different ways 😆)

ayushtues commented:
@yiyixuxu a quick suggestion, since the model was introduced as being a Multi-lingual version of text2img models, it might be a good idea to have some examples with prompts in other languages as well

yiyixuxu commented:
@patrickvonplaten @ayushtues 

Meet starry cat! (made with the cat image we generated + starry night) 

Image mixing is essentially just a text2img pipeline, but we use a mixture of images and texts to create image_embeddings - I think we can still pass a prompt, too, if we want. This is super cool, I think!  


```python
from diffusers import KandinskyPriorPipeline, KandinskyPipeline
from diffusers.utils import load_image
import PIL

import torch

from torchvision import transforms


def _convert_image_to_rgb(image):
    return image.convert("RGB")

image_transforms = transforms.Compose(
        [
            transforms.Resize(224, interpolation=transforms.InterpolationMode.BICUBIC),
            transforms.CenterCrop(224),
            _convert_image_to_rgb,
            transforms.ToTensor(),
            transforms.Normalize((0.48145466, 0.4578275, 0.40821073), (0.26862954, 0.26130258, 0.27577711)),
        ]
    )



REPO_PRIOR = "YiYiXu/Kandinsky-prior"
REPO = "YiYiXu/Kandinsky"

model_dtype = torch.float16

# create prior 
pipe_prior = KandinskyPriorPipeline.from_pretrained(REPO_PRIOR, torch_dtype=model_dtype)
pipe_prior.to("cuda")

def create_image_embeddings(
    pipe_prior, 
    conds, 
    weights, 
    num_inference_steps = 5,
    generator = None, 
    guidance_scale = 4.0,
    negative_prior_prompt = "",
    negative_prompt = "",):

    image_emb = []
    for cond, weight in zip(conds, weights):
        if isinstance(cond, str):
            # this is for testing only, normally we should pass it as argument
            generator = torch.Generator(device='cuda').manual_seed(0)
            emb = pipe_prior(
                cond, 
                num_inference_steps=num_inference_steps,
                generator=generator,
                guidance_scale=guidance_scale,
                negative_prompt=negative_prior_prompt).images
            
        elif isinstance(cond, PIL.Image.Image):
            cond = image_transforms(cond).unsqueeze(0).to(dtype = model_dtype, device="cuda")
            emb = pipe_prior.image_encoder(cond)["image_embeds"]
        
        else:
            raise ValueError(f"`cond` has to be of type `str` or `PIL.Image` but is {type(cond)}")
        
        image_emb.append(emb * weight)

    image_emb = torch.cat(image_emb).sum(dim=0, keepdim=True)

    zero_image_emb = pipe_prior(negative_prompt).images

    return image_emb, zero_image_emb

img1 = load_image("https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main" 
    "/kandinsky/cat.png")

img2 = load_image("https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main" 
    "/kandinsky/starry_night.jpeg")

images_texts = ["a cat", img1, img2 ]
weights = [0.3,0.3,0.4]

image_emb, zero_image_emb = create_image_embeddings(pipe_prior, images_texts, weights,)

pipe = KandinskyPipeline.from_pretrained(REPO, torch_dtype =model_dtype )
pipe.to("cuda")

generator = torch.Generator(device="cuda").manual_seed(0)
out = pipe(
    "",
    image_embeds=image_emb,
    negative_image_embeds =zero_image_emb,
    height=768,
    width=768,
    num_inference_steps=150,
    generator=generator )

image = out.images[0]
image.save("starry_cat.png")
```
![starry_cat](https://github.com/huggingface/diffusers/assets/12631849/b570408d-5bc7-4784-bbda-680005cd12ea)

patrickvonplaten commented:
@yiyixuxu I adapted the PR a bit removing the need for a whole text_proj model as stated [here](https://github.com/huggingface/diffusers/pull/3308#issuecomment-1543901510)

Updated your repos online accordingly: 
- https://huggingface.co/YiYiXu/Kandinsky-inpaint
- https://huggingface.co/YiYiXu/Kandinsky

and made sure that all the scripts that you posted above still work. Generations look great!

patrickvonplaten commented on src/diffusers/models/unet_2d_blocks.py:
```diff
@@ -2174,10 +2180,10 @@ def __init__(
-            hidden_states = resnet(hidden_states, temb=None)
-            hidden_states = attn(hidden_states)
+            hidden_states = resnet(hidden_states, temb=temb)
+            hidden_states = attn(hidden_states, vq_emb=temb)
```
```suggestion
            hidden_states = attn(hidden_states, temb=temb)
```

patrickvonplaten commented on src/diffusers/models/vae.py:
```diff
@@ -205,19 +209,23 @@ def __init__(
         self.gradient_checkpointing = False
 
-    def forward(self, z):
+    def forward(self, z, zq=None):
```
what does `zq` stand for?

patrickvonplaten commented on src/diffusers/models/vae.py:
```diff
@@ -205,19 +209,23 @@ def __init__(
         self.gradient_checkpointing = False
 
-    def forward(self, z):
+    def forward(self, z, zq=None):
```
```suggestion
    def forward(self, z, time_embed=None):
```

patrickvonplaten replied:
Would be in favor of renaming it to `latent_embeds` here

patrickvonplaten commented on src/diffusers/models/vq_model.py:
```diff
@@ -82,6 +82,7 @@ def __init__(
         norm_num_groups: int = 32,
         vq_embed_dim: Optional[int] = None,
         scaling_factor: float = 0.18215,
+        norm_type: str = "default",  # default, spatial
```
```suggestion
        norm_type: str = "group",  # group, spatial
```

patrickvonplaten commented on src/diffusers/models/attention_processor.py:
```diff
@@ -872,9 +882,19 @@ def __init__(self):
         residual = hidden_states
 
+        if attn.spatial_norm is not None:
+            hidden_states = attn.spatial_norm(hidden_states, vq_emb)
```
```suggestion
            hidden_states = attn.spatial_norm(hidden_states, temb)
```

patrickvonplaten commented on src/diffusers/models/attention_processor.py:
```diff
@@ -419,9 +425,13 @@ def __call__(
         hidden_states,
         encoder_hidden_states=None,
         attention_mask=None,
+        vq_emb=None,
```
```suggestion
        temb=None,
```

patrickvonplaten commented on src/diffusers/models/attention_processor.py:
```diff
@@ -419,9 +425,13 @@ def __call__(
         residual = hidden_states
 
+        if attn.spatial_norm is not None:
+            hidden_states = attn.spatial_norm(hidden_states, vq_emb)
```
```suggestion
            hidden_states = attn.spatial_norm(hidden_states, temb)
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,322 @@
+
+        batch_size = batch_size * num_images_per_prompt
+
+        if prompt == "" or prompt[0] == "":
+            image_embeddings = self.create_zero_img_emb(batch_size=batch_size, device=device)
+
+        else:
```
```suggestion
```
Let's actually completely remove this and let the user call `create_zero_img_emb` directly

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,322 @@
+        latents = latents * scheduler.init_noise_sigma
+        return latents
+
+    def create_zero_img_emb(self, batch_size, device):
```
```suggestion
    def get_zero_embed(self, batch_size=1, device=None):
        device = device or self.device
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,322 @@
+    @torch.no_grad()
+    def __call__(
+        self,
+        prompt,
```
```suggestion
        prompt: Union[str, List[str]],
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,322 @@
+
+            latents = self.prior.post_process_latents(latents)
+
+            image_embeddings = latents
```
```suggestion
            image_embeddings = latents
            zero_embeds = self.get_zero_embed(latents.shape[0], device=latents.device)
```

patrickvonplaten replied:
Computing this is **very** cheap, so I'd say we just always compute it?

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,322 @@
+            if not return_dict:
+                return (image_embeddings,)
+
+        return ImagePipelineOutput(images=image_embeddings)
```
```suggestion
        return ImagePipelineOutput(images=image_embeddings, zero_embeds=zeros_embeds)
```

patrickvonplaten commented on src/diffusers/schedulers/scheduling_unclip.py:
```diff
@@ -100,6 +100,15 @@ class UnCLIPScheduler(SchedulerMixin, ConfigMixin):
         prediction_type (`str`, default `epsilon`, optional):
             prediction type of the scheduler function, one of `epsilon` (predicting the noise of the diffusion process)
             or `sample` (directly predicting the noisy sample`)
+        thresholding (`bool`, default `False`):
```
Can we run the UnCLIP slow tests here to make sure we didn't break anything? Also @williamberman could you give this a look here?

patrickvonplaten commented on src/diffusers/schedulers/scheduling_unclip.py:
```diff
@@ -205,6 +261,8 @@ def step(
         generator=None,
         return_dict: bool = True,
+        # YiYi's TO-DO: batch_size argument for testing, remove this later
+        batch_size: Optional[int] = None,
```
yes let's not forget to remove this

yiyixuxu replied:
@patrickvonplaten I think it might be easier to keep this argument because our  `randn_tensor()` function https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py#L67 expects the `latent` tensor to be the same size as `batch_size`, and also the number of generators (if we pass a list of generators) 

if we are going to generate `latents` with `2 * batch_size` shape we will have to refactor the logic  - think it's unnecessary? 

let me know - if it's ok with you I will just remove this TO-DO comment

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py:
```diff
@@ -0,0 +1,388 @@
+            torch.cuda.empty_cache()  # otherwise we don't see the memory savings (but they probably exist)
+
+        hook = None
+        for cpu_offloaded_model in [self.text_encoder, self.unet, self.vae]:
```
```suggestion
        for cpu_offloaded_model in [self.text_encoder, self.unet, self.movq]:
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py:
```diff
@@ -0,0 +1,388 @@
+        height: int = 512,
+        width: int = 512,
+        num_inference_steps: int = 100,
+        guidance_scale: float = 4.0,
```
is that the recommended guidance_scale?

yiyixuxu replied:
@patrickvonplaten 

They don't have doc so there are no recommendations - their default is 7 - but 4 is what's used in their space. Would this be ok?

I also want to change the default `num_inference_steps` to `25` in prior pipeline - would that be ok? Their default is 5 but I found that it's not sufficient with prompts that are a little bit more complicated. `25` is a value that works great in most cases and that's what they used in their space too

patrickvonplaten replied:
Sure, let's change to `25` in the prior pipeline! Regarding the `guidance_scale`, up to you what you think works better!

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py:
```diff
@@ -0,0 +1,388 @@
+        image = image.clamp(0, 1)
+        image = image.cpu().permute(0, 2, 3, 1).float().numpy()
+
+        if output_type == "pil":
```
Let's make sure to allow all `output_type == "pil"` or `output_type == "pt"` or `output_type == "np"` , but only those (so let's add ability to return PT and also add a test that checks that only pt np and pil are allowed)

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_inpaint.py:
```diff
@@ -0,0 +1,572 @@
+            torch.cuda.empty_cache()  # otherwise we don't see the memory savings (but they probably exist)
+
+        hook = None
+        for cpu_offloaded_model in [self.text_encoder, self.unet, self.vae]:
```
```suggestion
        for cpu_offloaded_model in [self.text_encoder, self.unet, self.movq]:
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_inpaint.py:
```diff
@@ -0,0 +1,572 @@
+        image = image.clamp(0, 1)
+        image = image.cpu().permute(0, 2, 3, 1).float().numpy()
+
+        if output_type == "pil":
```
Let's make sure to allow all `output_type == "pil"` or `output_type == "pt"` or `output_type == "np"` , but only those (so let's add ability to return PT and also add a test that checks that only pt np and pil are allowed)

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,322 @@
+        guidance_scale: float = 4.0,
+        output_type: Optional[str] = "pt",  # pt only
+        return_dict: bool = True,
+    ):
```
Need docstrings

patrickvonplaten commented on tests/pipelines/kandinsky/test_kandinsky_inpaint.py:
```diff
@@ -0,0 +1,318 @@
+            intermediate_size=37,
+            num_attention_heads=4,
+            num_hidden_layers=5,
+            vocab_size=250002,
```
This will create a huge vocab size vector, can we try to reduce this to just `1000`?

patrickvonplaten replied:
```suggestion
            vocab_size=1005,
```
See comments for `test_kadinsky`

patrickvonplaten commented on tests/pipelines/kandinsky/test_kandinsky_img2img.py:
```diff
@@ -0,0 +1,324 @@
+            intermediate_size=37,
+            num_attention_heads=4,
+            num_hidden_layers=5,
+            vocab_size=250002,
```
```suggestion
            vocab_size=1000,
```

To big for fast test

patrickvonplaten replied:
See comments for `test_kadinsky`

patrickvonplaten commented on tests/pipelines/kandinsky/test_kandinsky.py:
```diff
@@ -0,0 +1,291 @@
+            intermediate_size=37,
+            num_attention_heads=4,
+            num_hidden_layers=5,
+            vocab_size=250002,
```
```suggestion
            vocab_size=1000,
```

To big for fast test

patrickvonplaten commented on tests/pipelines/kandinsky/test_kandinsky.py:
```diff
@@ -0,0 +1,291 @@
+    # YiYi's TO-DO: add a tiny tokenizer?
+    @property
+    def dummy_tokenizer(self):
+        tokenizer = XLMRobertaTokenizer.from_pretrained("YiYiXu/Kandinsky", subfolder="tokenizer")
```
```suggestion
        tokenizer = XLMRobertaTokenizerFast.from_pretrained("YiYiXu/tiny-random-mclip-base")
```
added a small tokenizer for you there with 1005 vocab size

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,45 @@
+
+As text and image encoder it uses CLIP model and diffusion image prior (mapping) between latent spaces of CLIP modalities. This approach increases the visual performance of the model and unveils new horizons in blending images and text-guided image manipulation.
+
+The Kandinsky model in diffusers comes from ai-forever and the original codebase can be found [here](https://github.com/ai-forever/Kandinsky-2)
```
Let's add some cool use cases / examples here

sayakpaul replied:
```suggestion
The Kandinsky model in diffusers comes from `ai-forever` and the original codebase can be found [here](https://github.com/ai-forever/Kandinsky-2).
```

yiyixuxu replied:
@patrickvonplaten is this mention ok here?

yiyixuxu commented:
@patrickvonplaten do you have an updated conversions script?  

I was testing with the [decoder_fp16.ckpt](https://huggingface.co/ai-forever/Kandinsky_2.1/blob/main/decoder_fp16.ckpt) but it was in my to-do to update it with the decoder.ckpt - but I saw you updated the unet, maybe it was used with the decoder.ckpt already? 

I think I addressed all the feedbacks - except the cool examples in doc which is still WIP. I will update the slow test once confirm unet checkpoint is the final one

sayakpaul commented:
> I think I addressed all the feedbacks - except the cool examples in doc which is still WIP. 

I think we can work on a separate doc to discuss the capabilities of the model in details so that the community benefits from it. Happy to give a hand there :)

patrickvonplaten commented:
> I was testing with the [decoder_fp16.ckpt](https://huggingface.co/ai-forever/Kandinsky_2.1/blob/main/decoder_fp16.ckpt) but it was in my to-do to update it with the decoder.ckpt - but I saw you updated the unet, maybe it was used with the decoder.ckpt already?

I actually just did it manually by hand, so didn't look into the conversion scripts, but should be quite easy to update :-)

patrickvonplaten commented on tests/pipelines/kandinsky/test_kandinsky_img2img.py:
```diff
@@ -0,0 +1,315 @@
+        image_slice = image[0, -3:, -3:, -1]
+        image_from_tuple_slice = image_from_tuple[0, -3:, -3:, -1]
+
+        print(f"image.shape {image.shape}")
```
```suggestion
```

patrickvonplaten commented on tests/pipelines/kandinsky/test_kandinsky_inpaint.py:
```diff
@@ -0,0 +1,308 @@
+from ..test_pipelines_common import PipelineTesterMixin, assert_mean_pixel_difference
+
+
+torch.backends.cuda.matmul.allow_tf32 = False
+torch.use_deterministic_algorithms(True)
```
```suggestion
enable_full_determinism()
```

patrickvonplaten commented on tests/pipelines/kandinsky/test_kandinsky_inpaint.py:
```diff
@@ -0,0 +1,308 @@
+from diffusers import KandinskyInpaintPipeline, KandinskyPriorPipeline, UnCLIPScheduler, UNet2DConditionModel, VQModel
+from diffusers.pipelines.kandinsky.text_encoder import MCLIPConfig, MultilingualCLIP
+from diffusers.utils import floats_tensor, load_image, load_numpy, slow, torch_device
+from diffusers.utils.testing_utils import require_torch_gpu
```
```suggestion
from diffusers.utils.testing_utils import require_torch_gpu, enable_full_determinism
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,576 @@
+        zero_embeds = self.get_zero_embed(latents.shape[0], device=latents.device)
+
+        if output_type not in ["pt", "np"]:
+            deprecation_message = (
+                f"the output_type {output_type} is outdated and has been set to `np`. Please make sure to set it to one of these instead: "
+                "`np`, `pt``"
+            )
+            deprecate("Unsupported output_type", "1.0.0", deprecation_message, standard_warn=False)
+            output_type = "np"
```
```suggestion
            raise ValueError(f"Only the output types "pt" and "np" are supported not `output_type={output_type}`")
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/text_encoder.py:
```diff
@@ -0,0 +1,32 @@
+        embs = self.transformer(input_ids=input_ids, attention_mask=attention_mask)[0]
+        embs2 = (embs * attention_mask.unsqueeze(2)).sum(dim=1) / attention_mask.sum(dim=1)[:, None]
+        return self.LinearTransformation(embs2), embs
+
+    @classmethod
+    def _load_state_dict_into_model(cls, model, state_dict, pretrained_model_name_or_path, _fast_init=True):
+        model.load_state_dict(state_dict)
+        return model, [], [], []
```
We don't need this I think

patrickvonplaten commented:
As discussed offline, let's try to use DDIM scheduler for both img2img and text2img

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py:
```diff
@@ -0,0 +1,468 @@
+            ):
+                noise_pred, _ = noise_pred.split(latents.shape[1], dim=1)
+
+            if i + 1 == timesteps_tensor.shape[0]:
+                pass
+            else:
+                timesteps_tensor[i + 1]
```
```suggestion
            if i + 1 == timesteps_tensor.shape[0]:
                pass
            else:
                timesteps_tensor[i + 1]
```
Is this the intended design? Should ```timesteps_tensor[i + 1]``` be ```t = timesteps_tensor[i + 1]````?

yiyixuxu replied:
@patrickvonplaten 

oh sorry. this step was to create the `prev_timestep` that was needed for unclip scheduler and it was no longer needed for DDIM or DDPM. it was something like before and `make style` automatically changed it to this
```
if i + 1 == timesteps_tensor.shape[0]:
                prev_timestep = None
else:
                prev_timestep = timesteps_tensor[i + 1]
```


Will delete and will remember to check more thoroughly after `make style` in the future

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+
+Kandinsky 2.1 inherits best practicies from Dall-E 2 and Latent diffusion, while introducing some new ideas.
+
+As text and image encoder it uses CLIP model and diffusion image prior (mapping) between latent spaces of CLIP modalities. This approach increases the visual performance of the model and unveils new horizons in blending images and text-guided image manipulation.
```
```suggestion
Kandinsky 2.1 inherits best practices from [DALL-E 2](https://arxiv.org/abs/2204.06125) and [Latent Diffusion](https://arxiv.org/abs/2112.10752), while introducing some new ideas.
```

sayakpaul replied:
```suggestion
It uses [CLIP](https://huggingface.co/docs/transformers/model_doc/clip) for encoding images and text, and a diffusion image prior (mapping) between latent spaces of CLIP modalities. This approach enhances the visual performance of the model and unveils new horizons in blending images and text-guided image manipulation.
```

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+
+## Overview
+
+Kandinsky 2.1 inherits best practicies from Dall-E 2 and Latent diffusion, while introducing some new ideas.
```
```suggestion
Kandinsky 2.1 inherits best practices from [DALL-E 2](https://arxiv.org/abs/2204.06125) and [Latent Diffusion](https://arxiv.org/abs/2112.10752), while introducing some new ideas.
```

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+
+## Usage example
+
+In the following, we will walk you through some cool examples of using Kandinsky pipelines to create incredible artwork beyond imagination.
```
```suggestion
In the following, we will walk you through some cool examples of using the Kandinsky pipelines to create some visually aesthetic artwork.
```

The previous was sounding a bit too affirmative. I think it makes sense to not do that in the documentation. Let the users be the judge of that :)

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+
+### Text-to-Image Generation
+
+For text-to-image generation, we need to use both [`KandinskyPriorPipeline] and [`KandinskyPipeline`]. The first step is to create CLIP image embedding conditioned on a text prompt. Let's throw a fun prompt at Kandinsky to see what it will come up with :)
```
```suggestion
For text-to-image generation, we need to use both [`KandinskyPriorPipeline`] and [`KandinskyPipeline`]. The first step is to create CLIP image embedding conditioned on a text prompt. Let's throw a fun prompt at Kandinsky to see what it comes up with :)
```

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
````diff
@@ -0,0 +1,285 @@
+negative_prompt = "low quality, bad quality"
+```
+
+You should create the CLIP image embeddings for both your `prompt` and `negavie_prompt`. Also, the prior pipeline we use to create these CLIP image embeddings is a diffusion model itself, so you can use `negative_prompt`, `guidance_scale`, and `num_inference_steps` arguments to guide this process, just like how you would normally do with all other pipelines in diffusers.
````
As a user, this is a bit confusing to me -- creating image embeddings from prompts (i.e., text content). Worth clarifying this bit I think.

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+pipe_prior = KandinskyPriorPipeline.from_pretrained("YiYiXu/Kandinsky-prior", torch_dtype=torch.float16)
+pipe_prior.to("cuda")
+
+generator = torch.Generator(device="cuda").manual_seed(12)
+image_emb = pipe_prior(
+    prompt, guidance_scale=1.0, num_inference_steps=25, generator=generator, negative_prompt=negative_prompt
+).images
+
+zero_image_emb = pipe_prior(
+    negative_prompt, guidance_scale=1.0, num_inference_steps=25, generator=generator, negative_prompt=negative_prompt
+).images
```
Consider discussing why the `guidance_scale` is so low.

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+
+![img](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/kandinsky-docs/cheeseburger.png)
+
+Kandinsky model works really extremely well with creative prompts. Here is some of the amazing art we created using exact same process but different prompts.
```
```suggestion
The Kandinsky model works extremely well with creative prompts. Here is some of the amazing art we created using the exact same process but with different prompts.
```

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
````diff
@@ -0,0 +1,285 @@
+prompt = "birds eye view of a quilted paper style alien planet landscape, vibrant colours, Cinematic lighting"
+```
+![img](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/kandinsky-docs/alienplanet.png)
+
````
If you have more examples, feel free to log them on Weights and Biases and leave a link here. Optional, though.

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+The same Kandinsky model weights can be used for text-guided image-to-image translation. In this case, just make sure to load the weights using the [`KandinskyImg2ImgPipeline`] pipeline.
+
+**Note**: You can also directly move the weights of the text-to-image pipelines to the image-to-image pipelines
+without loading them twice by making use of the [`~DiffusionPipeline.components()`] function as explained [here](#converting-between-different-pipelines).
```
```suggestion
without loading them twice by making use of the [`~DiffusionPipeline.components`] function as explained [here](#converting-between-different-pipelines).
```

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+
+One cheeseburger monster coming up! Enjoy! 
+
+![img](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/kandinsky-docs/cheeseburger.png)
```
The image paths need to be corrected it seems?

yiyixuxu replied:
corrected!

sayakpaul commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+    "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main" "/kandinsky/cat.png"
+)
+
+mask = np.ones((768, 768), dtype=np.float32)
+mask[:250, 250:-250] = 0
```
Maybe add a comment before this block?

yiyixuxu commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+from diffusers import KandinskyPipeline, KandinskyPriorPipeline
+import torch
+
+# create prior
```
yep yep once an agreement is made :)

yiyixuxu replied:
updated!

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+
+### Text-to-Image Generation
+
+For text-to-image generation, we need to use both [`KandinskyPriorPipeline`] and [`KandinskyPipeline`]. The first step is to create CLIP image embedding conditioned on a text prompt. Let's throw a fun prompt at Kandinsky to see what it comes up with :)
```
```suggestion
For text-to-image generation, we need to use both [`KandinskyPriorPipeline`] and [`KandinskyPipeline`]. The first step is to encode text prompts with CLIP and then diffuse the CLIP text embeddings to CLIP image embeddings, as first proposed in [DALL-E 2](https://cdn.openai.com/papers/dall-e-2.pdf). Let's throw a fun prompt at Kandinsky to see what it comes up with :)
```

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
````diff
@@ -0,0 +1,285 @@
+negative_prompt = "low quality, bad quality"
+```
+
+You should create the CLIP image embeddings for both your `prompt` and `negavie_prompt`. Also, the prior pipeline we use to create these CLIP image embeddings is a diffusion model itself, so you can use `negative_prompt`, `guidance_scale`, and `num_inference_steps` arguments to guide this process, just like how you would normally do with all other pipelines in diffusers.
````
```suggestion
We will pass both the `prompt` and `negative_prompt` to our prior diffusion pipeline. In contrast to other diffusion pipelines, such as Stable Diffusion, the `prompt` and `negative_prompt` shall be passed separately so that we can retrieve a CLIP image embedding for each prompt input. You can use `guidance_scale`, and `num_inference_steps` arguments to guide this process, just like how you would normally do with all other pipelines in diffusers.
```

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
````diff
@@ -0,0 +1,285 @@
+You should create the CLIP image embeddings for both your `prompt` and `negavie_prompt`. Also, the prior pipeline we use to create these CLIP image embeddings is a diffusion model itself, so you can use `negative_prompt`, `guidance_scale`, and `num_inference_steps` arguments to guide this process, just like how you would normally do with all other pipelines in diffusers.
+
+```python
+from diffusers import KandinskyPipeline, KandinskyPriorPipeline
````
```suggestion
from diffusers import KandinskyPriorPipeline
```

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
````diff
@@ -0,0 +1,285 @@
+Once we create the image embedding, we can use [`KandinskyPipeline`] to generate images.
+
+```python
+from PIL import Image
````
```suggestion
from PIL import Image
from diffusers import KandinskyPipeline
```

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+)
+
+mask = np.ones((768, 768), dtype=np.float32)
+mask[:250, 250:-250] = 0
```
```suggestion
# Let's mask out an area above the cat's head
mask[:250, 250:-250] = 0
```

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+	- all
+	- __call__
+
+[[autodoc]] KandinskyInpaintPipeline
```
```suggestion
## KandinskyInpaintPipeline

[[autodoc]] KandinskyInpaintPipeline
```

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+
+[[autodoc]] KandinskyInpaintPipeline
+	- all
+	- __call__
```
```suggestion
	- __call__
	
## KandinskyImg2ImgPipeline
```

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+	- all
+	- __call__
+
+[[autodoc]] KandinskyPriorPipeline
+	- all
+	- __call__
+	- interpolate
```
```suggestion
```

Let's move this up to be the 1st pipeline

patrickvonplaten replied:
https://github.com/huggingface/diffusers/pull/3308/files#r1203866574

patrickvonplaten commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+![img](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/kandinsky-docs/starry_cat.png)
+
+
+## KandinskyPipeline
```
```suggestion
## KandinskyPriorPipeline

[[autodoc]] KandinskyPriorPipeline
	- all
	- __call__
	- interpolate
	
## KandinskyPipeline
```

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky.py:
```diff
@@ -0,0 +1,463 @@
+        >>> from diffusers import KandinskyPipeline, KandinskyPriorPipeline
+        >>> import torch
+
+        >>> pipe_prior = KandinskyPriorPipeline.from_pretrained("YiYiXu/Kandinsky-prior")
```
```suggestion
        >>> pipe_prior = KandinskyPriorPipeline.from_pretrained("kandinsky-community/Kandinsky-prior")
```

And all other fields

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,570 @@
+import numpy as np
+import PIL
+import torch
+from torchvision import transforms
```
```suggestion
```

**Careful here!** We don't have a required dependency on `torchvision` for `diffusers`. Please make sure to remove the dependency here

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,570 @@
+
+image_transforms = transforms.Compose(
+    [
+        transforms.Resize(224, interpolation=transforms.InterpolationMode.BICUBIC),
+        transforms.CenterCrop(224),
+        _convert_image_to_rgb,
+        transforms.ToTensor(),
+        transforms.Normalize((0.48145466, 0.4578275, 0.40821073), (0.26862954, 0.26130258, 0.27577711)),
```
```suggestion
        transforms.Resize(224, interpolation=transforms.InterpolationMode.BICUBIC),
        transforms.CenterCrop(224),
        _convert_image_to_rgb,
        transforms.ToTensor(),
        transforms.Normalize((0.48145466, 0.4578275, 0.40821073), (0.26862954, 0.26130258, 0.27577711)),
```

All these transforms can be done without using `torchvision`

patrickvonplaten commented on src/diffusers/pipelines/kandinsky/pipeline_kandinsky_prior.py:
```diff
@@ -0,0 +1,570 @@
+        Examples:
+
+        Returns:
+            `tuple`
```
```suggestion
            [`KandinskyPriorPipelineOutput`] or `tuple`
```

patrickvonplaten reviewed (commented):
Great we're almost done now! Left some more comments. Two bigger refactors we still need to do:

1.) Remove the dependency on `torchvision`. This is **very** important as otherwise `diffusers` will be broken
2.) Let's revert all changes to unclip scheduler as we don't need it anymore. Think it's more interesting to get the other schedulers (DPM etc...) working with Kandinsky instead.

yiyixuxu commented on docs/source/en/api/pipelines/kandinsky.mdx:
```diff
@@ -0,0 +1,285 @@
+    "https://huggingface.co/datasets/hf-internal-testing/diffusers-images/resolve/main" "/kandinsky/starry_night.jpeg"
+)
+
+images_texts = ["a cat", img1, img2]
```
done

patrickvonplaten commented:
Good to merge!

patrickvonplaten commented:
We should try training ControlNet on it!
